table/resizing.go

```diff
@@ -0,0 +1,401 @@
+					biggestDiffToMedian = diffToMedian
+					biggestDiffToMedianIndex = j
+				}
+			}
+
+			if biggestDiffToMedianIndex <= 0 || colWidths[biggestDiffToMedianIndex] == 0 {
+				break
+			}
+			colWidths[biggestDiffToMedianIndex]--
+		}
+	}
+
+	shrinkBiggestColumns(true)
+	shrinkToMedian()
+	shrinkBiggestColumns(false)
+
+	return colWidths, r.expandRowHeigths(colWidths)
+}
+
+// expandRowHeigths expands the row heights.
+func (r *resizer) expandRowHeigths(colWidths []int) (rowHeights []int) {
+	rowHeights = r.defaultRowHeigths()
+	if !r.wrap {
+		return rowHeights
+	}
+	for i, row := range r.allRows {
+		for j, cell := range row {
+			height := r.detectContentHeight(cell, colWidths[j]-r.xPaddingForCol(j)) + r.xPaddingForCell(i, j)
+			if height > rowHeights[i] {
+				rowHeights[i] = height
+			}
+		}
+	}
+	return
+}
+
+// defaultRowHeigths returns the default row heights.
+func (r *resizer) defaultRowHeigths() (rowHeights []int) {
+	rowHeights = make([]int, len(r.allRows))
+	for i := range rowHeights {
+		if i < len(r.rowHeights) {
+			rowHeights[i] = r.rowHeights[i]
+		}
+		if rowHeights[i] < 1 {
+			rowHeights[i] = 1
+		}
+	}
+	return
+}
+
+// maxColumnWidths returns the maximum column widths.
+func (r *resizer) maxColumnWidths() []int {
+	maxColumnWidths := make([]int, len(r.columns))
+	for i, col := range r.columns {
+		if col.fixedWidth > 0 {
+			maxColumnWidths[i] = col.fixedWidth
+		} else {
+			maxColumnWidths[i] = col.max + r.xPaddingForCol(col.index)
+		}
+	}
+	return maxColumnWidths
+}
+
+// columnCount returns the column count.
+func (r *resizer) columnCount() int {
+	return len(r.columns)
+}
+
+// maxCharCount returns the maximum character count.
+func (r *resizer) maxCharCount() int {
+	var count int
+	for _, col := range r.columns {
+		if col.fixedWidth > 0 {
+			count += col.fixedWidth - r.xPaddingForCol(col.index)
+		} else {
+			count += col.max
+		}
+	}
+	return count
+}
+
+// maxTotal returns the maximum total width.
+func (r *resizer) maxTotal() (maxTotal int) {
+	for j, column := range r.columns {
+		if column.fixedWidth > 0 {
+			maxTotal += column.fixedWidth
+		} else {
+			maxTotal += column.max + r.xPaddingForCol(j)
+		}
+	}
+	return
+}
+
+// totalPadding returns the total padding.
+func (r *resizer) totalPadding() (totalPadding int) {
+	for _, col := range r.columns {
+		totalPadding += col.xPadding
+	}
+	return
+}
+
+// xPaddingForCol returns the horizontal padding for a column.
+func (r *resizer) xPaddingForCol(j int) int {
+	if j >= len(r.columns) {
+		return 0
+	}
+	return r.columns[j].xPadding
+}
+
+// xPaddingForCell returns the horizontal padding for a cell.
+func (r *resizer) xPaddingForCell(i, j int) int {
+	if i >= len(r.yPaddings) || j >= len(r.yPaddings[i]) {
+		return 0
+	}
+	return r.yPaddings[i][j]
+}
+
+// totalBorder returns the total border.
+func (r *resizer) totalBorder() int {
+	return (r.columnCount() * r.borderPerCell()) + r.extraBorder()
```

Comment:
is this only getting the horizontal borders inside the table? I.e. not counting the outlining border widths of the table?

Naming-wise we might want to make it more similarly to the `Style` methods e.g. `GetHorizontalBorderSize` and `GetVerticalBorderSize`

Reply:
It's counting the total number of border chars in a row.

```
| foo | bar | baz |
```

This numbers tends to be the number of columns + 1. For 3 cells, we have 4 chars for border.

Reply:
Regarding the naming, I slightly changing the name of these functions in my last commit.